docs/shared/extending-nx/project-inference-plugins.md

```diff
@@ -0,0 +1,100 @@
+# Project Inference
+
+> This API is experimental and might change.
+
+Project inference describes the ability of Nx to discover and work with projects based on source code and configuration files in your repo.
+
+One of the best examples of this in Nx is that `package.json` scripts are automatically inferred as targets for your workspace. If using nx core, and no plugins, you can also see this as projects defined in your `package.json` workspace are inferred as Nx projects.
+
+## Adding Plugins to Workspace
+
+You can register a plugin by adding it to the plugins array in `nx.json`:
+
+```json
+{
+  ...,
+  "plugins": [
+    "awesome-plugin"
+  ]
+}
+```
+
+## Project File Patterns
+
+Project file patterns are used in two scenarios:
+
+- Inferring projects when `workspace.json` is not present
+- Determining which files should be passed into `registerProjectTargets`.
+
+Lets use the below plugin and workspace layout as an example:
+
+> libs/awesome-plugin/index.ts
+
+```typescript
+export const projectFilePatterns = ['project.json', 'my-other-project-file'];
+export function registerProjectTargets(projectFilePath) {
+  console.log(projectFilePath);
+}
+```
+
+> workspace layout
+
+```text
+my-workspace/
+├─ node_modules/
+├─ libs/
+│  ├─ my-project/
+│  │  ├─ my-other-project-file
```

Comment:
Yeah, this is currently the case. Thats only the behavior when taking advantage of full project inference though, when just using register project targets the execution order is a bit more abstract.

Under the hood, when running a target this is the path Nx takes:
1. Read project configurations
2. Check that target exists
  a. If not, check if an inference plugin would register that target. If so, register it.
3. Run the target

Since reading in project configurations is not coupled to the actual target inference, you can get some of the other behavior.